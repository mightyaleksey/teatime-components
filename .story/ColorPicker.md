## Preview


### size s


```
<div>
  some text for the baseline
  <ColorPicker
    defaultValue='CCCCFF'
    name='palette'
    size='s'/>
  {' '}
  <ColorPicker
    defaultValue='CCCCFF'
    disabled={true}
    name='palette'
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <ColorPicker
    name='palette'
    size='m'/>
  {' '}
  <ColorPicker
    disabled={true}
    name='palette'
    size='m'/>
</div>
```


### size l

```
<div>
  some text for the baseline
  <ColorPicker
    defaultValue='FFCC99'
    name='palette'
    size='l'/>
  {' '}
  <ColorPicker
    defaultValue='FFCC99'
    disabled={true}
    name='palette'
    size='l'/>
</div>
```
