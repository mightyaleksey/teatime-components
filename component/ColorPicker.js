'use strict';

const {Component, PropTypes} = require('react');
const {TAB} = require('../lib/keyCode');
const {cssColorValue, userColorValue} = require('../lib/color');
const {filterProps, isControlled, genericName, nullToString} = require('../lib/util');
const {isUndefined, map, noop} = require('../lib/dash');
const Overlay = require('../view/Overlay');
const React = require('react');
const Tile = require('../view/Tile');
const cc = require('classnames');

const cssModules = {
  l: require('../style/colorPicker/colorPicker-l.css'),
  m: require('../style/colorPicker/colorPicker-m.css'),
  s: require('../style/colorPicker/colorPicker-s.css'),
};

class ColorPicker extends Component {
  constructor(props) {
    super(props);

    this._controlled = isControlled(props);

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      isOpened: false,
      value: isUndefined(value) ? '' : nullToString(userColorValue(value)),
    };
  }

  componentWillReceiveProps(nextProps) {
    this._controlled = isControlled(nextProps);

    this.setState({
      value: this._controlled
        ? nullToString(nextProps.value)
        : this.state.value,
    });
  }

  css = tokenName => genericName(this.props, tokenName)

  focus() {
    if (!this._input) return;
    this._input.focus();
  }

  select() {
    if (!this._input) return;
    this._input.select();
  }

  _onChange = e => {
    const value = e.target.value;

    if (!this._controlled) this.setState({isOpened: false, value});
    this.props.onChange(e, {value});
  }

  _onClearClick = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      value: '',
    });

    this.focus();
  }

  _onClick = () => {
    this.setState({isOpened: !this.state.isOpened});
  }

  _onInputBlur = e => {
    const {value} = this.state;
    const nextValue = userColorValue(value);

    if (nextValue === value) return;

    if (!this._controlled) this.setState({
      isOpened: false,
      value: nextValue,
    });

    this.props.onChange(e, {value: nextValue});
  }

  _onInputFocus = () => {
    if (!this.state.isOpened) return;
    this.setState({isOpened: false});
  }

  _onKeyDown = ({keyCode}) => {
    if (this.props.disabled) return;
    if (keyCode === TAB) this.setState({isOpened: false});
  }

  _onTileClick = (e, value) => {
    if (!this._controlled) this.setState({isOpened: false, value});
    this.props.onChange(e, {value});
    this.focus();
  }

  _onOutsideClick = () => {
    if (!this.state.isOpened) return;
    this.setState({isOpened: false});
  }

  _parentNode = () => this.refs.parent

  computeTiles(palette) {
    if (!this.state.isOpened) return null;

    const {css} = this;

    const menuItem = css('menuItem');

    return map(color =>
      this.renderTile({
        className: menuItem,
        color,
        key: color,
        onClick: this._onTileClick,
      }), palette);
  }

  render() {
    const {className} = this.props;
    const {css} = this;

    return (
      <span
        {...filterProps(this.props)}
        className={cc(css('container'), className)}
        onChange={void 0}
        ref='parent'>
        {this.renderButton()}
        {this.renderClear()}
        {this.renderInput()}
        {this.renderMenu()}
      </span>
    );
  }

  renderButton() {
    const {disabled} = this.props;
    const {value} = this.state;
    const {css} = this;

    return (
      <button
        className={css('button')}
        disabled={disabled}
        onClick={this._onClick}
        style={{
          backgroundColor: cssColorValue(value),
        }}/>
    );
  }

  renderClear() {
    const {disabled} = this.props;
    const {value} = this.state;

    if (disabled || !value) return null;

    const {css} = this;

    return (
      <span
        className={css('clear')}
        disabled={disabled}
        onClick={this._onClearClick}
        value={value}/>
    );
  }

  renderInput() {
    const {autoFocus, disabled, id, name, placeholder} = this.props;
    const {value} = this.state;
    const {css} = this;

    return (
      <input
        autoFocus={autoFocus}
        className={css('control')}
        disabled={disabled}
        id={id}
        name={name}
        onBlur={this._onInputBlur}
        onChange={this._onChange}
        onFocus={this._onInputFocus}
        placeholder={placeholder}
        ref={r => this._input = r}
        type='text'
        value={value}/>
    );
  }

  renderMenu() {
    const {palette} = this.props;
    const {isOpened} = this.state;
    const {css} = this;

    return (
      <Overlay
        className={cc(css('menu'), {
          [css('isClosedMenu')]: !isOpened,
        })}
        onOutsideClick={this._onOutsideClick}
        parentNode={this._parentNode}>
        {this.computeTiles(palette)}
      </Overlay>
    );
  }

  renderTile(tileProps) {
    return (
      <Tile {...tileProps}/>
    );
  }
}

ColorPicker.defaultProps = {
  onChange: noop,
  palette: [
    '000000', 'CC0000', 'CC6600',
    'CCCC00', '66CC00', '00CC00',
    '00CC66', '00CCCC', '0066CC',
    '0000CC', '6600CC', 'CC00CC',
    'CC0066', '333333', 'FF0000',
    'FF8000', 'FFFF00', '80FF00',
    '00FF00', '00FF80', '00FFFF',
    '007FFF', '0000FF', '7F00FF',
    'FF00FF', 'FF0080', '666666',
    'FF3333', 'FF9933', 'FFFF33',
    '99FF33', '33FF33', '33FF99',
    '33FFFF', '3399FF', '3333FF',
    '9933FF', 'FF33FF', 'FF3399',
    '999999', 'FF6666', 'FFB366',
    'FFFF66', 'B3FF66', '66FF66',
    '66FFB3', '66FFFF', '66B2FF',
    '6666FF', 'B266FF', 'FF66FF',
    'FF66B3', 'CCCCCC', 'FF9999',
    'FFCC99', 'FFFF99', 'CCFF99',
    '99FF99', '99FFCC', '99FFFF',
    '99CCFF', '9999FF', 'CC99FF',
    'FF99FF', 'FF99CC', 'FFFFFF',
    'FFCCCC', 'FFE6CC', 'FFFFCC',
    'E6FFCC', 'CCFFCC', 'CCFFE6',
    'CCFFFF', 'CCE5FF', 'CCCCFF',
    'E5CCFF', 'FFCCFF', 'FFCCE6',
  ],
  size: 'm',
  styles: cssModules,
};

ColorPicker.propTypes = {
  autoFocus: PropTypes.bool,
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  palette: PropTypes.arrayOf(PropTypes.string),
  placeholder: PropTypes.string,
  size: PropTypes.oneOf([
    'l',
    'm',
    's',
  ]),
  styles: PropTypes.object,
  value: PropTypes.string,
};

module.exports = ColorPicker;
