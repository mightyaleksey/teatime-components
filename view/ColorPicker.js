'use strict';

const { Component, PropTypes } = require('react');
const { bind, hasValueProp, sanitizeProps } = require('../tool/component');
const { classNames, styleName } = require('../tool/className');
const { isUndefined, noop } = require('../tool/func');
const { isHexBased, normalizeColor, trimHash } = require('../tool/color');
const Input = require('./Input');
const Overlay = require('./Overlay');
const Tile = require('./Tile');
const React = require('react');
const reactOutsideEvent = require('../mixin/ReactOutsideEvent');
const warning = require('../tool/warning');

var didWarnForDefaultValue = false;

class ColorPicker extends Component {
  constructor(props) {
    super(props);

    bind(this, [
      'onChange',
      'onInputBlur',
      'onInputFocus',
      'onKeyDown',
      'onPreviewClick',
      'onTileClick',
    ]);

    this.controlled = hasValueProp(props);

    if (process.env.NODE_ENV !== 'production' && this.controlled && !didWarnForDefaultValue) { // eslint-disable-line no-undef
      warning(typeof props.defaultValue === 'undefined',
        'ColorPicker contains an input of type text with both value and defaultValue props. ' +
        'Input elements must be either controlled or uncontrolled ' +
        '(specify either the value prop, or the defaultValue prop, but not ' +
        'both). Decide between using a controlled or uncontrolled input ' +
        'element and remove one of these props. More info: ' +
        'https://fb.me/react-controlled-components');

      didWarnForDefaultValue = true;
    }

    var value = this.controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      isOpened: false,
      value: !isUndefined(value)
        ? value
        : 'FFFFFF',
    };
  }

  componentWillReceiveProps(nextProps) {
    if (this.controlled) {
      this.setState({value: nextProps.value});
    }
  }

  focus() {
    if (this.refs.control) {
      this.refs.control.focus();
    }
  }

  onChange(e, data) {
    this.updateValue(e, data.value);
  }

  onInputBlur(e) {
    var nextValue = normalizeColor(this.state.value);
    if (nextValue === this.state.value) return;
    this.updateValue(e, normalizeColor(nextValue || 'FFFFFF'));
  }

  onInputFocus() {
    // @todo add possibility to control focus from outside
    if (this.state.isOpened === false) return;
    this.setState({isOpened: false});
  }

  onKeyDown(e) {
    if (this.props.disabled) return;

    const preview = this.refs.preview;

    if (e.keyCode === 9 && preview && preview === e.target) { // tab
      this.setState({isOpened: false});
    }
  }

  onPreviewClick() {
    this.setState({isOpened: !this.state.isOpened});
  }

  onOutsideEvent() {
    if (this.state.isOpened === false) return;
    this.setState({isOpened: false});
  }

  onTileClick(e, color) {
    this.updateValue(e, trimHash(color));
    this.focus();
  }

  updateValue(e, value) {
    if (!this.controlled) {
      this.setState({isOpened: false, value});
    }

    this.props.onChange(e, {isOpened: false, value});
  }

  render() {
    const { id, styles, ...o } = this.props;

    return (
      <div
        className={styleName(this.props, {isFixedWrapper: this.props.hasFixedWidth})}
        onKeyDown={this.onKeyDown}>
        {this.renderPreview()}
        <Input
          {...sanitizeProps(['hasFixedWidth', 'palette'], o)}
          className={undefined}
          defaultValue={undefined}
          id={id}
          onBlur={this.onInputBlur}
          onChange={this.onChange}
          onFocus={this.onInputFocus}
          ref='control'
          styleName={undefined}
          styles={styles}
          value={this.state.value}/>
        {this.renderMenu()}
      </div>
    );
  }

  renderPreview() {
    const { disabled, styles } = this.props;
    const value = this.state.value;

    return (
      <button
        className={styles.preview}
        disabled={disabled}
        onClick={this.onPreviewClick}
        ref='preview'
        style={{
          background: `#${isHexBased(value) ? trimHash(value) : 'fff'}`,
        }}/>
    );
  }

  renderMenu() {
    const { styles } = this.props;

    return (
      <Overlay className={classNames(styles.menu, {
        [styles.isClosedMenu]: !this.state.isOpened,
        [styles.isOpenedMenu]: this.state.isOpened,
        [styles.isFixedMenu]: this.props.hasFixedWidth,
      })}>
        {this.renderTiles()}
      </Overlay>
    );
  }

  renderTiles() {
    if (!this.state.isOpened) {
      return null;
    }

    const { palette, styles } = this.props;
    const tiles = [];

    for (var length = palette.length, i = 0; i < length; ++i) {
      const tile = palette[i];

      tiles.push((
        <Tile
          color={`#${tile}`}
          key={`_${tile}`}
          onClick={this.onTileClick}
          styles={styles}/>
      ));
    }

    return tiles;
  }
}

ColorPicker.defaultProps = {
  hasFixedWidth: false,
  onChange: noop,
  palette: [
    '000000',
    'CC0000',
    'CC6600',
    'CCCC00',
    '66CC00',
    '00CC00',
    '00CC66',
    '00CCCC',
    '0066CC',
    '0000CC',
    '6600CC',
    'CC00CC',
    'CC0066',
    '333333',
    'FF0000',
    'FF8000',
    'FFFF00',
    '80FF00',
    '00FF00',
    '00FF80',
    '00FFFF',
    '007FFF',
    '0000FF',
    '7F00FF',
    'FF00FF',
    'FF0080',
    '666666',
    'FF3333',
    'FF9933',
    'FFFF33',
    '99FF33',
    '33FF33',
    '33FF99',
    '33FFFF',
    '3399FF',
    '3333FF',
    '9933FF',
    'FF33FF',
    'FF3399',
    '999999',
    'FF6666',
    'FFB366',
    'FFFF66',
    'B3FF66',
    '66FF66',
    '66FFB3',
    '66FFFF',
    '66B2FF',
    '6666FF',
    'B266FF',
    'FF66FF',
    'FF66B3',
    'CCCCCC',
    'FF9999',
    'FFCC99',
    'FFFF99',
    'CCFF99',
    '99FF99',
    '99FFCC',
    '99FFFF',
    '99CCFF',
    '9999FF',
    'CC99FF',
    'FF99FF',
    'FF99CC',
    'FFFFFF',
    'FFCCCC',
    'FFE6CC',
    'FFFFCC',
    'E6FFCC',
    'CCFFCC',
    'CCFFE6',
    'CCFFFF',
    'CCE5FF',
    'CCCCFF',
    'E5CCFF',
    'FFCCFF',
    'FFCCE6',
  ],
  styleName: 'container',
};

// @todo add color validation for value prop
ColorPicker.propTypes = {
  hasFixedWidth: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  palette: PropTypes.array,
  styleName: PropTypes.string,
  styles: PropTypes.object,
};

module.exports = reactOutsideEvent(ColorPicker, ['click']);
