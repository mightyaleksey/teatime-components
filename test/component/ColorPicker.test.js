import ColorPicker from '../../component/ColorPicker';
import React from 'react';
import renderer from 'react-test-renderer';

test('ColorPicker renders correctly with defaults', () => {
  const tree = renderer.create(
    <ColorPicker
      name='colorPicker control'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('ColorPicker renders correctly with defined props', () => {
  const tree = renderer.create(
    <ColorPicker
      autoFocus={true}
      disabled={true}
      id='colorPickers id'
      name='colorPicker control'
      placeholder='FFFFFF'
      value='#bebebe'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
