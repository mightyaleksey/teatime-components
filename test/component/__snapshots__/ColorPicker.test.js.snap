exports[`test ColorPicker renders correctly with defaults 1`] = `
<span
  className="container"
  onChange={undefined}>
  <button
    className="button"
    disabled={undefined}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
      }
    } />
  <input
    autoFocus={undefined}
    className="control"
    disabled={undefined}
    id={undefined}
    name="colorPicker control"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    type="text"
    value="" />
  <div
    className="menu isClosedMenu" />
</span>
`;

exports[`test ColorPicker renders correctly with defined props 1`] = `
<span
  className="container"
  onChange={undefined}>
  <button
    className="button"
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#bebebe",
      }
    } />
  <input
    autoFocus={true}
    className="control"
    disabled={true}
    id="colorPickers id"
    name="colorPicker control"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="FFFFFF"
    type="text"
    value="BEBEBE" />
  <div
    className="menu isClosedMenu" />
</span>
`;
